File: python/1/historian_hysteria.py
from collections import Counter

def main():
    # Read input
    left_list = []
    right_list = []
    try:
        while True:
            line = input()
            if not line.strip():
                continue  # Skip empty lines
            tokens = line.strip().split()
            if len(tokens) != 2:
                raise ValueError(f"Invalid input line: {line}")
            left_num = int(tokens[0])
            right_num = int(tokens[1])
            left_list.append(left_num)
            right_list.append(right_num)
    except EOFError:
        pass

    # Part One
    total_distance = calculate_total_distance(left_list, right_list)
    print(f"Total distance: {total_distance}")

    # Part Two
    similarity_score = calculate_similarity_score(left_list, right_list)
    print(f"Similarity score: {similarity_score}")

def calculate_total_distance(left_list, right_list):
    """
    Calculate the total distance between the sorted left and right lists.
    """
    sorted_left = sorted(left_list)
    sorted_right = sorted(right_list)
    distances = [abs(a - b) for a, b in zip(sorted_left, sorted_right)]
    total_distance = sum(distances)
    return total_distance

def calculate_similarity_score(left_list, right_list):
    """
    Calculate the similarity score based on the frequency of numbers
    from the left list appearing in the right list.
    """
    right_counts = Counter(right_list)
    similarity_score = sum(num * right_counts.get(num, 0) for num in left_list)
    return similarity_score

if __name__ == "__main__":
    main()

File: rust/1/historian_hysteria.rs
use std::collections::HashMap;
use std::io::{self, BufRead};

fn main() {
    // Read input
    let (left_list, right_list) = read_input();

    // Part One: Calculate Total Distance
    let total_distance = calculate_total_distance(&left_list, &right_list);
    println!("Total distance: {}", total_distance);

    // Part Two: Calculate Similarity Score
    let similarity_score = calculate_similarity_score(&left_list, &right_list);
    println!("Similarity score: {}", similarity_score);
}

fn read_input() -> (Vec<i32>, Vec<i32>) {
    let stdin = io::stdin();
    let mut left_list = Vec::new();
    let mut right_list = Vec::new();

    for line in stdin.lock().lines() {
        let line = line.expect("Failed to read line");
        if line.trim().is_empty() {
            continue; // Skip empty lines
        }
        let tokens: Vec<&str> = line.trim().split_whitespace().collect();
        if tokens.len() != 2 {
            panic!("Invalid input line: {}", line);
        }
        let left_num = tokens[0].parse::<i32>().expect("Invalid number");
        let right_num = tokens[1].parse::<i32>().expect("Invalid number");
        left_list.push(left_num);
        right_list.push(right_num);
    }

    (left_list, right_list)
}

fn calculate_total_distance(left_list: &Vec<i32>, right_list: &Vec<i32>) -> i32 {
    let mut sorted_left = left_list.clone();
    let mut sorted_right = right_list.clone();

    sorted_left.sort();
    sorted_right.sort();

    let distances: Vec<i32> = sorted_left
        .iter()
        .zip(sorted_right.iter())
        .map(|(a, b)| (a - b).abs())
        .collect();

    distances.iter().sum()
}

fn calculate_similarity_score(left_list: &Vec<i32>, right_list: &Vec<i32>) -> i32 {
    let mut right_counts: HashMap<i32, i32> = HashMap::new();

    for &num in right_list {
        *right_counts.entry(num).or_insert(0) += 1;
    }

    left_list
        .iter()
        .map(|&num| num * right_counts.get(&num).unwrap_or(&0))
        .sum()
}